### PR TITLE
Minimal loop + DTW: update data_loader; add smoke test

### PR DESCRIPTION
**Pull request by ImYangYun**
This PR adds a minimal training/validation loop that uses masked MSE and a validation-time DTW sanity metric.

- The loop keeps tensors in [B,T,J,C] for loss/metrics and only permutes to [B,J,C,T] at the model boundary.
- The data loader now filters by split, uses a safe pivot in [1, total_frames-1] so past/future are non-empty, and returns masked tensors.
- A small smoke test (with a DummyModel) runs training_step and validation_step without heavy dependencies.
- I ran the minimal loop locally for ~500–1000 steps, it trains and validates successfully, and the curves look stable. The smoke test also passes.

And I met some problems when implementing these tasks:
In the loop I convert BTJC → BJCT before calling the model, and then convert the model output BJCT → BTJC after the forward pass. The model was implemented around BJCT (time last), while our masks, zero_pad_collator, and loss/metrics (masked MSE + DTW) are written for BTJC (time at dim=1). So inputs/past are permuted to match the model, and predictions are permuted back so losses operate over [B, T, J, C].
I’m not sure if this is the best place to keep the conversions. Right now it works, but the double permute feels a bit clunky. Can you give some suggestions on this? Thanks!

minimal_curves

### CODE REVIEW

File: signwriting_animation/scripts/minimal_loop.py [hunk 1]
```diff
@@ -0,0 +1,265 @@
...
+    """
+    Build a DataLoader with:
+    - Dataset returning MaskedTensor.
+    - zero_pad_collator to align sequences along time dimension.
+    - FilteredDataset to pick a small, valid subset for a minimal run.
+    """
+    base = DynamicPosePredictionDataset(data_dir=data_dir, csv_path=csv_path, with_metadata=True, split=split)
+    ds = FilteredDataset(base, target_count=200, max_scan=3000)
+    print(f"[DEBUG] split={split} | batch_size={bs} | len(ds)={len(ds)}")
+    return DataLoader(ds, batch_size=bs, shuffle=False, num_workers=num_workers, collate_fn=zero_pad_collator)
```
Comment (AmitMY) [inline signwriting_animation/scripts/minimal_loop.py:235]: you probably want `shuffle=true` for train

File: signwriting_animation/scripts/minimal_loop.py [hunk 2]
```diff
@@ -0,0 +1,265 @@
...
+        return self.base[self.idx[i]]
+
+class LitMinimal(pl.LightningModule):
+    """
+    Minimal Lightning module:
+    - Forward: SignWritingToPoseDiffusion (expects BJCT)
+    - Loss: masked MSE; plus DTW in validation as a sanity metric
+    - No checkpointing; meant for quick end-to-end checks.
+    """
+    def __init__(self, num_keypoints=586, num_dims=3, lr=1e-3, log_dir="logs"):
```
Comment (AmitMY) [inline signwriting_animation/scripts/minimal_loop.py:129]: for testing purposes, i recommend using `reduce_holistic` to reduce the number of keypoints.
you could also use `normalize_mean_std` to normalize the data distribution https://github.com/sign-language-processing/pose-anonymization/blob/main/pose_anonymization/data/normalization.py#L65-L76

File: signwriting_animation/scripts/minimal_loop.py [hunk 3]
```diff
@@ -0,0 +1,265 @@
...
+                        torch.minimum(Cmat[i-1, j], Cmat[i, j-1]),
+                        Cmat[i-1, j-1]
+                    )
+                Cmat[i, j] = D[i, j] + m
+
+        vals.append(Cmat[t-1, t-1] / (2 * t))
+
+    if not vals:
+        return torch.tensor(0.0, device=pred.device)
+    return torch.stack(vals).mean()
```
Comment (AmitMY) [inline signwriting_animation/scripts/minimal_loop.py:95]: we have metrics for poses in this repo: https://github.com/sign-language-processing/pose-evaluation/
including some examples in [here](https://github.com/sign-language-processing/pose-evaluation/blob/main/pose_evaluation/examples/example_metric_construction.py) - feel free to open an issue if you are struggling

File: signwriting_animation/scripts/minimal_loop.py [hunk 4]
```diff
@@ -0,0 +1,265 @@
...
+            mask = (mask.abs().sum(dim=(2, 3)) > 0).float()
+        elif mask.dim() == 3:
+            mask = (mask.abs().sum(dim=2) > 0).float()
+        sign_img = cond["sign_image"].float()
+
+        x_bjct = btjc_to_bjct(fut)
+        past_bjct = btjc_to_bjct(past)
+        timesteps = torch.zeros(x_bjct.size(0), dtype=torch.long, device=x_bjct.device)
+        out = self.forward(x_bjct, timesteps, past_bjct, sign_img)
+        pred = bjct_to_btjc(out)
```
Comment (AmitMY) [inline signwriting_animation/scripts/minimal_loop.py:159]: if you move the `btjc_to_bjct` and `bjct_to_btjc` inside the forward call, that would be less code duplication, and we know the forward expects this.

File: signwriting_animation/scripts/minimal_loop.py [hunk 5]
```diff
@@ -0,0 +1,265 @@
...
+        if not self.idx:
+            self.idx = [0]
+
+    def __len__(self):
+        return len(self.idx)
+
+    def __getitem__(self, i):
+        return self.base[self.idx[i]]
+
+class LitMinimal(pl.LightningModule):
```
Comment (AmitMY) [inline signwriting_animation/scripts/minimal_loop.py:122]: sure, this is minimal, but ideally, you continue developing it. It should be `SignWritingToPoseDiffusionLightning` and you can use it minimally with arguments - for example if you want no checkpointing, tell the `Trainer` to not checkpoint, etc.

Comment (AmitMY) [inline signwriting_animation/scripts/minimal_loop.py:122]: generally, this class should not be in "scripts", it can be `lightning_module.py` or something.

File: signwriting_animation/scripts/minimal_loop.py [hunk 6]
```diff
@@ -0,0 +1,265 @@
...
+        devices=1,
+        log_every_n_steps=5,
+        limit_train_batches=10,
+        limit_val_batches=5,
+        check_val_every_n_epoch=1,
+        enable_checkpointing=False,
+        deterministic=True,
+        num_sanity_val_steps=0,
+    )
+    trainer.fit(model, train_loader, val_loader)
```
Comment (AmitMY) [inline signwriting_animation/scripts/minimal_loop.py:265]: ok, let's say it fits. now what? how can we validate that it is correct? that the model can reproduce a single training example? you can train on like 4 examples from the dataset (the loss will go down fast), and then take these samples, and run the inference loop - given the first frame, predict the next `future_frames` etc. then you will see with your eyes if the model is able to overfit or not

**Review by AmitMY — CHANGES_REQUESTED** [review]
looks ok, left a few comments.

**Comment by ImYangYun** [comment]
### Changes
- Update model with PositionalEncoding + TimestepEmbedder
- Achieves stable motion generation (disp_ratio: 0.30 → 1.02)
- Fixed pylint warnings and improved code quality

### Results (4-sample test)
- Displacement Ratio: 1.02 
- MPJPE: 0.019
- PCK@0.1: 100%

train_curve

**Comment by ImYangYun** [comment]
Fixed lint, suppressed R0913 and R0917 in pyproject.toml and added ignore-paths for translation.
